### PR TITLE
rls: rls request factory (aka key builder map)

### PR DESCRIPTION
**Pull request by creamsoup**

### CODE REVIEW

File: rls/src/main/java/io/grpc/rls/internal/RlsRequestFactory.java [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+import io.grpc.rls.internal.RlsProtoData.NameMatcher;
+import io.grpc.rls.internal.RlsProtoData.RouteLookupConfig;
+import io.grpc.rls.internal.RlsProtoData.RouteLookupRequest;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import javax.annotation.CheckReturnValue;
+
+/**
+ * A RlsRequestFactory creates {@link RouteLookupRequest} using key builder map from {@link
```
[the remark this answers is not in the document]
Comment (creamsoup) [reply rls/src/main/java/io/grpc/rls/internal/RlsRequestFactory.java:37]: i read it a route lookup service request factory. so both are right depends on how you pronounce RLS.

File: rls/src/main/java/io/grpc/rls/internal/RlsRequestFactory.java [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+              path,
+              nameMatcher.getKey(),
+              new NameMatcher(nameMatcher.getKey(), requestHeaders, nameMatcher.isOptional()));
+        }
+      }
+    }
+    return table;
+  }
+
+  /** Creates an {@link RouteLookupRequest} for given request's metadata. */
```
[the remark this answers is not in the document]
Comment (creamsoup) [reply rls/src/main/java/io/grpc/rls/internal/RlsRequestFactory.java]: this one a is correct =p

File: rls/src/main/java/io/grpc/rls/internal/RlsRequestFactory.java [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+        List<String> requestHeaders = nameMatcher.names();
+        for (Name name : grpcKeyBuilder.getNames()) {
+          String method =
+              name.getMethod() == null || name.getMethod().isEmpty()
+                  ? "*" : name.getMethod();
+          String path = name.getService() + "/" + method;
+          table.put(
+              path,
+              nameMatcher.getKey(),
+              new NameMatcher(nameMatcher.getKey(), requestHeaders, nameMatcher.isOptional()));
```
[the remark this answers is not in the document]
Comment (creamsoup) [reply rls/src/main/java/io/grpc/rls/internal/RlsRequestFactory.java]: yes right! it used to ignore the isOptional but not anymore.

File: rls/src/main/java/io/grpc/rls/internal/RlsRequestFactory.java [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+  }
+
+  @Override
+  public String toString() {
+    return MoreObjects.toStringHelper(this)
+        .add("lookupService", target)
+        .add("keyBuilderTable", keyBuilderTable)
+        .toString();
+  }
+}
```
[the remark this answers is not in the document]
Comment (creamsoup) [reply rls/src/main/java/io/grpc/rls/internal/RlsRequestFactory.java]: done.

File: rls/src/main/java/io/grpc/rls/internal/RlsRequestFactory.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (dapengzhang0) [reply rls/src/main/java/io/grpc/rls/internal/RlsRequestFactory.java:37]: Well, not a big deal.

**Pull request merged**